evm/src/modules/TokenGateway.sol

```diff
@@ -169,92 +157,112 @@ contract TokenGateway is BaseIsmpModule {
 
-        if (erc20 != address(0) && !params.redeem) {
-            require(IERC20(erc20).transferFrom(from, address(this), params.amount), "Insufficient user balance");
+        if (erc20 != address(0) && !teleportParams.redeem) {
+            require(IERC20(erc20).transferFrom(from, address(this), teleportParams.amount), "Insufficient user balance");
         } else if (erc6160 != address(0)) {
-            IERC6160Ext20(erc6160).burn(from, params.amount, "");
+            IERC6160Ext20(erc6160).burn(from, teleportParams.amount, "");
         } else {
             revert("Unknown Token Identifier");
         }
 
+        uint256 fee = (IIsmpHost(_params.host).perByteFee() * data.length) + teleportParams.fee;
         // only swap if the feeToken is not the token intended for fee
-        if (feeToken != params.feeToken) {
-            // Calculate output fee in the fee token before swap:
-            uint256 fee = (IIsmpHost(_params.host).perByteFee() * data.length + 1) + params.fee;
-
+        if (feeToken != teleportParams.feeToken) {
             address[] memory path = new address[](2);
             // from
-            path[0] = params.feeToken;
+            path[0] = teleportParams.feeToken;
             // to
             path[1] = feeToken;
 
             require(
-                IERC20(params.feeToken).transferFrom(from, address(this), params.amountInMax),
+                IERC20(teleportParams.feeToken).transferFrom(from, address(this), teleportParams.amountInMax),
                 "insufficient funds for intended fee token"
             );
-            require(IERC20(params.feeToken).approve(_params.uniswapV2, params.amountInMax), "approve failed");
+            require(
+                IERC20(teleportParams.feeToken).approve(_params.uniswapV2, teleportParams.amountInMax), "approve failed"
+            );
             IUniswapV2Router(_params.uniswapV2).swapTokensForExactTokens(
-                fee, params.amountInMax, path, from, block.timestamp
+                fee, teleportParams.amountInMax, path, address(this), block.timestamp
             );
+        } else {
+            require(IERC20(feeToken).transferFrom(from, address(this), fee), "user has insufficient funds");
         }
 
-        // permit the host with the exact amount
-        IAllowanceTransfer(_params.permit2).permit(from, params.hostPermit.permit, params.hostPermit.signature);
+        // approve the host with the exact amount
+        require(IERC20(feeToken).approve(_params.host, fee), "approve failed");
         DispatchPost memory request = DispatchPost({
-            dest: params.dest,
+            dest: teleportParams.dest,
             to: abi.encodePacked(address(this)),
-            body: bytes.concat(hex"00", data), // add enum variant for body
-            timeout: params.timeout,
-            fee: params.fee,
+            body: data,
+            timeout: teleportParams.timeout,
+            fee: teleportParams.fee,
             payer: msg.sender
```

Comment:
Was thinking the payer for this dispatch call to the host would be the application in this case `Gateway`, [address(this)]

Reply:
The payer is the account to be refunded in the case of a timeout